docs/Developer Documentation/Logger/index.md

```diff
@@ -0,0 +1,159 @@
+
+
+SDL implements Logger abstration for easily replacement of logging library.
+
+|||
+High Level Design
+![Logger](./assets/high_level_design.png)
+|||
+
+
+All components have aceess to `Logger` interface and use it for logging.
+
+### Logger interface 
+
+Logger macroses use Logger interface for sending messages to External Logger. 
+Logger interface contains only methods required by any SDL component to perform logging : 
+
+ * instance() - singleton access
+ * PushLog(LogMessage)
+ * IsEnabledFor(LogLevel)
+ * DeInit()
+ * Flush() 
+
+
+## LoggerImpl
+
+`Logger` interface is implemented by `LoggerImpl`. 
+
+`LoggerImpl` use message loop thread to proxy log messages to third party (external) logger.
+`LoggerImpl` owns `ThirdPartyLoggerInterface` and controls it's lifetime. 
+`LoggerImpl` provides implementation of the singleton pattern.
+
+
+### Message loop thread in SDLLogger
+
+Message loop thread is needed to avoid significant performance degradation in run time as logging calls are blocking calls and might take too much time. 
+`LoggerImpl::PushLog` is non a blocking call. It will put the log message into the queue and returns immediately 
+
+
+If `ThirdPartyLoggerInterface` supports non blocking threaded logging, minor changes in `LoggerImpl` can be created with `use_message_loop_thread = false`. 
+
+## Logger singleton 
+
+Logger is the only one singleton class in SDL.
+Singleton pattern required to have an access to logger from any component. 
+`Logger::instance()` provides singleton by `Logger` interface.
+So SDL components do not have information about neither logger implementation nor specific external logger. 
+
+## Logger singleton with plugins 
+
+SDL plugins are shared libraries, so `LoggerSingleton` could not be implemented with Mayers singleton. 
+Mayers singleton would create own SDL logger instance for each plugin.
+
+The idea is to pass singleton pointer to Plugin during creation, so that plugin could initialize `Logger::instance` pointer with one received from SDL core. 
+
+
+Singleton Instance implementation : 
+```cpp
+// ilogger.h
+static Logger& instance(Logger* pre_init = nullptr);
+
+...
+// logger_impl.cc
+Logger& Logger::instance(Logger* pre_init) {
+  static Logger* instance_ = nullptr;
+  if (pre_init) {
+    assert(instance_ == nullptr);
+    instance_ = pre_init;
+  }
+  assert(instance_);
+  return *instance_;
+}
+```
+
+`pre_init` is `nullptr` by default, so all components will access instance_ static pointer for logging. 
+`main()`  function need to create `SDLLogger` implementation and call `Logger::instance` (logger implementation);
+
+Plugin implementation:
+```cpp 
+extern "C" PluginType* Create(Logger* logger_singleton_instance) {
+  Logger::instance(logger_instance);
+  return new PluginType();
+}
+```
+
+SDL Core part will pass pointer to logger singleton to the plugin so that plugin shared lib could initialize `Logger::instance` with the same pointer as core part. 
```

Comment:
SDL Core will pass a pointer to the logger singleton to the plugin so that the plugin shared lib can initialize `Logger::instance` with the same pointer as the core portion.

Reply:
Fixed with  
aeadebe